lib/util/progress.coffee

```diff
@@ -1,53 +1,238 @@
+    @stack.push p
+    @emitter.emit 'did-update-stack'
+
+  delete: (p) ->
+    @activate()
+    i = @stack.findIndex (e) -> e.id is p.id
+    return if i < 0
+    @stack.splice i, 1
+    @emitter.emit 'did-update-stack'
+
+  update: (p) ->
+    @activate()
+    i = @stack.findIndex (e) -> e.id is p.id
+    return if i < 0
+    oldp = @stack[i]
+    @stack[i] = p
+    @emitter.emit 'did-update-progress', p
+    # if determinate-ness changes, update the stack display
+    @emitter.emit 'did-update-stack' unless oldp.progress*p.progress > 0
+
+  emptyStack: () ->
+    @stack = []
+    @emitter.emit 'did-update-stack'
+
+  emptyProgress: (id = 'empty') ->
+    id: id
+    progress: 0
+
+  indeterminateProgress: (id = 'indeterminate') ->
+    id: id
+    progress: -1
+
+  # update logic
+  hasNoDeterminateBars: ->
+    @stack.filter((p) => p.progress >= 0).length is 0
+
+  onDidUpdateStack: (f) -> @emitter.on 'did-update-stack', f
+
+  onDidUpdateProgress: (f) -> @emitter.on 'did-update-progress', f
+
+  # UI elements:
   progressView: (p) ->
     span = document.createElement 'span'
-    span.classList.add 'inline-block'
+    span.id = "ink-prog-#{p.id}"
     prog = document.createElement 'progress'
     prog.classList.add 'ink'
     prog.setAttribute 'max', 1
-    if p.progress? then prog.setAttribute 'value', p.progress
     span.appendChild prog
 
-    Object.observe p, (changes) ->
-      for {name, object} in changes
-        continue unless name == "progress"
-        if object.progress?
-          prog.setAttribute 'value', object.progress
-        else
-          prog.removeAttribute 'value'
+    updateView = (prg) =>
+      return unless p.id is prg.id
+      if prg.progress >= 0
+        prog.setAttribute 'value', prg.progress
+      else
+        prog.removeAttribute 'value'
+
+    @onDidUpdateProgress (prg) => updateView(prg)
+    updateView p
 
     span
 
-  showTile: (p) ->
-    @tile?.destroy()
-    @tile = @statusBar?.addLeftTile
-      item: @progressView(p)
-      priority: -1
+  msgView: (p, parent) ->
+    div = document.createElement 'div'
+    div.classList.add 'ink-tooltip-msg'
+    div.appendChild.innerText = p.msg
+
+    @onDidUpdateProgress (prg) =>
+      return unless p.id is prg.id
+      if prg.msg?.length
+        div.innerText = prg.msg
+        parent.classList.add 'has-tooltip'
+      else
+        parent.classList.remove 'has-tooltip'
+
+    div
+
+  tableRowView: (p) ->
+    tr = document.createElement 'tr'
+    # left text
+    tdl = document.createElement 'td'
+    tdl.classList.add 'progress-tr'
+    p.leftText? and tdl.appendChild document.createTextNode p.leftText
+    # progress bar
+    td = document.createElement 'td'
+    td.classList.add 'progress-tr', 'has-tooltip'
+    td.appendChild @progressView p
+    td.appendChild @msgView p, td
+    # right text
+    tdr = document.createElement 'td'
+    tdr.classList.add 'progress-tr'
+    p.rightText? and tdr.appendChild document.createTextNode p.rightText
```

Comment:
I'm quite interested in [Etch](https://github.com/atom/etch) by the Atom devs. I think there's quite a few places in our code base that it could clean up a lot.